### PR TITLE
.Net: Bentho/lazy step initialization

### PR DESCRIPTION
**Pull request by alliscode**
### Description

This PR contains the following changes:

#### First commit: 

Refactoring initialization of steps within a local process run lazily and only once. The previous behavior would instantiate and initialize all of steps, including calling their state initialize method, all at once before the process started running. After this change a step is not instantiated and initialized until right before it receives its first event. This change also ensures that a step will not be initialized twice.

#### Second commit:

Similar to the first but this time applied to the process initialization.

#### Third commit:

Adding more unit tests for LocalProcess.cs

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Review by crickman — APPROVED** [review]

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 1]
```diff
@@ -39,47 +40,37 @@ internal class LocalStep : KernelProcessMessageChannel
...
-        this.Id = id;
+        this.Id = stepInfo.State.Id ?? Guid.NewGuid().ToString("n");
+        this.Name = stepInfo.State.Name;
         this.ParentProcessId = parentProcessId;
         this.LoggerFactory = loggerFactory;
         this._kernel = kernel;
+        this._stepInfo = stepInfo;
+        this._initializeTask = new Lazy<ValueTask>(this.InitializeStepAsync);
+        this._logger = this.LoggerFactory?.CreateLogger(this._stepInfo.InnerStepType) ?? new NullLogger<LocalStep>();
+        this._outputEdges = this._stepInfo.Edges.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToList());
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:61]: should there be a check before the dictionary to ensure that there are not repeated outputEdges names? prob they will get overwritten

**Review by esttenorio — APPROVED** [review]

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs [hunk 2]
```diff
@@ -56,7 +57,7 @@ public LocalProcess(KernelProcess process, Kernel kernel, string? parentProcessI
     /// Loads the process and initializes the steps. Once this is complete the process can be started.
     /// </summary>
     /// <returns>A <see cref="Task"/></returns>
-    public async Task LoadAsync()
+    public ValueTask InitializeProcessAsync()
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs:60]: Does this need to be public not that the initialization is done via `_initialiseTask`?

If it stays public how do we prevent initialization being performed multiple times?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs [hunk 3]
```diff
@@ -109,12 +110,14 @@ public async Task LoadAsync()
...
+        // Lazy one-time initialization of the process before staring it.
+        await this._initializeTask.Value.ConfigureAwait(false);
+
         this._processCancelSource = new CancellationTokenSource();
         this._processTask = this._joinableTaskFactory.RunAsync(()
             => this.Internal_ExecuteAsync(kernel, keepAlive: keepAlive, cancellationToken: this._processCancelSource.Token));
-        return Task.CompletedTask;
     }
 
     public async Task RunOnceAsync(KernelProcessEvent? processEvent, Kernel? kernel = null)
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs:123]: Missing XML documentation

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 4]
```diff
@@ -39,47 +40,37 @@ internal class LocalStep : KernelProcessMessageChannel
...
     /// <param name="parentProcessId">Optional. The Id of the parent process if one exists.</param>
     /// <param name="loggerFactory">An instance of <see cref="LoggerFactory"/> used to create loggers.</param>
-    public LocalStep(string name, string id, Kernel kernel, string? parentProcessId = null, ILoggerFactory? loggerFactory = null)
+    public LocalStep(KernelProcessStepInfo stepInfo, Kernel kernel, string? parentProcessId = null, ILoggerFactory? loggerFactory = null)
     {
-        Verify.NotNullOrWhiteSpace(name);
-        Verify.NotNullOrWhiteSpace(id);
+        Verify.NotNull(stepInfo);
         Verify.NotNull(kernel);
+        Verify.NotNullOrWhiteSpace(stepInfo.State.Name);
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:51]: Should verify of name happen when state is created?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 5]
```diff
@@ -39,47 +40,37 @@ internal class LocalStep : KernelProcessMessageChannel
...
 
-        this.Name = name;
-        this.Id = id;
+        this.Id = stepInfo.State.Id ?? Guid.NewGuid().ToString("n");
+        this.Name = stepInfo.State.Name;
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:54]: A reference to `stepInfo` is being kept so `Name` property can point to `this._stepInfo.State.Name`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 6]
```diff
@@ -39,47 +40,37 @@ internal class LocalStep : KernelProcessMessageChannel
...
     {
-        Verify.NotNullOrWhiteSpace(name);
-        Verify.NotNullOrWhiteSpace(id);
+        Verify.NotNull(stepInfo);
         Verify.NotNull(kernel);
+        Verify.NotNullOrWhiteSpace(stepInfo.State.Name);
 
-        this.Name = name;
-        this.Id = id;
+        this.Id = stepInfo.State.Id ?? Guid.NewGuid().ToString("n");
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:53]: Will be problematic if this step has a different id to the `stepInfo.State`?

Also is `stepInfo.State.Id` being null valid?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 7]
```diff
@@ -39,47 +40,37 @@ internal class LocalStep : KernelProcessMessageChannel
...
     }
 
     /// <summary>
     /// Initializes the step with the provided step information.
     /// </summary>
-    /// <param name="stepInfo">An instance of <see cref="KernelProcessStepInfo"/></param>
     /// <returns>A <see cref="ValueTask"/></returns>
     /// <exception cref="KernelException"></exception>
-    public virtual async ValueTask InitializeAsync(KernelProcessStepInfo stepInfo)
+    internal virtual async ValueTask InitializeStepAsync()
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:69]: Can this be made private if initialization should happen via `this._initializeTask`?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs (hunk 1 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:61]: You are correct. Currently the process only supports in memory state persistence so things will get reset when a process is recreated. Refactoring this out will come soon in the work to define the state management that will be common to the runtimes.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs (hunk 2 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs:60]: The class is internal, but your point is valid. Made this private.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs (hunk 6 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs]: Good catch, fixed.

**Review by markwallace-microsoft — APPROVED** [review]

**Review by esttenorio — APPROVED** [review]

**Review by crickman — APPROVED** [review]

**Pull request merged**